### ISSUE TITLE
Explanation of nn.LookupTable Behavior

### ISSUE THREAD

**Issue by stevenfortney** [Element-Research/rnn#268]
This may be a very basic question but I am trying to understand what the use of nn.LookupTable is in the examples for this repo. Nearly every example has something of the sort:

`-- Recurrence.recurrentModule
local rm = nn.Sequential()
  :add(nn.ParallelTable()
      :add(nn.LookupTable(nIndex, hiddenSize)) 
      :add(nn.Linear(hiddenSize, hiddenSize))) 
   :add(nn.CAddTable())
   :add(nn.Sigmoid())
  :add(nn.FastLSTM(hiddenSize,hiddenSize)) -- an AbstractRecurrent instance
  :add(nn.Linear(hiddenSize,hiddenSize))
  :add(nn.Sigmoid())  `

This is from the [(https://github.com/Element-Research/rnn/blob/master/examples/nested-recurrence-lstm.lua)] example. Basically, you sequentially feed the RNN a set of _1 X batchSize_ inputs, rho times, but the LookupTable has nIndex=10 as default. How does this nIndex relate to the inputs? And what does the LookupTable do to the inputs? Thanks.

**Comment by nicholas-leonard** [comment]
> Basically, you sequentially feed the RNN a set of 1 X batchSize inputs, rho times, but the LookupTable has nIndex=10 as default. How does this nIndex relate to the inputs?
Say you have a vocabulary of 10 words and you want to learn a vector representing each word. The vectors share a common embedding space. That is what you use the lookup table for. So my input might be 3 words, each indexing one of the 10 rows of the table: `[4,3,2]`. After forwarding this input of size 3 through the lookup table, the output will be a matrix where each row is the commensurate embedding vector of the word (so 3 rows).

> Also as a last comment, it seems that the LookupTabel is unnecesary to begin with because the input vectors are numeric. What do you gain from encoding them in a LookupTable?
You get a learned embedding space for the words.
